### ISSUE TITLE
Check for minimum python version 2.7.3

### ISSUE THREAD

**Issue by cbm755** [cbm755/octsympy#843]
On help@octave mailing list, Anoop has reported failures that seem to suggest we don't play nice with Windows and Python 2.  This is strange because the bundle package uses Python 2 and that has been tested...

- - - - - - - -

Based on your detailed output (thanks!), python appears to be working correctly.  It looks like some kind of newline problem...  Considering the "A" output before the error message: it looks like each string has a extraneous newline appended.  The "out" variable looks wrong too: on my system it has things like "<f>Communication established.</f>" but on Anoop's machine, there are extra newlines.

My first reaction was no one has tested the combination of Windows and Python2.  But that seems unlikely; the wiki [1] explicitly mentions Python 2.

https://github.com/cbm755/octsympy/wiki/Notes-on-Windows-installation

- - - - - - - -

Here is the relevant output:
````
GNU Octave, version 4.2.1
Copyright (C) 2017 John W. Eaton and others.
This is free software; see the source code for copying conditions.
There is ABSOLUTELY NO WARRANTY; not even for MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  For details, type 'warranty'.

Octave was configured for "x86_64-w64-mingw32".

Additional information about Octave is available at http://www.octave.org.

Please contribute if you find this software useful.
For more information, visit http://www.octave.org/get-involved.html

Read http://www.octave.org/bugs.html to learn how to submit bug reports.
For information about changes from previous versions, type 'news'.

>> pkg load symbolic
>> syms x
OctSymPy v2.6.0: this is free software without warranty, see source.
Initializing communication with SymPy using a popen2() pipe.
Some output from the Python subprocess (pid 9436) might appear next.
Python 2.7 (r27:82525, Jul  4 2010, 07:43:08) [MSC v.1500 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> >>>
A =
{
  [1,1] =
Communication established.

  [1,2] =
1.1.1

  [1,3] =
2.7 (r27:82525, Jul  4 2010, 07:43:08) [MSC v.1500 64 bit (AMD64)]

}
out = <output_block>
<list>
<item>
<f>
1003
</f>
<f>
Communication established.
</f>
</item>
<item>
<f>
1003
</f>
<f>
1.1.1
</f>
</item>
<item>
<f>
1003
</f>
<f>
2.7 (r27:82525, Jul  4 2010, 07:43:08) [MSC v.1500 64 bit (AMD64)]
</f>
</item>
</list>
</output_block>
Closing the Python pipe...
error: ipc_popen2: something unexpected has gone wrong in starting python
error: called from
    python_ipc_popen2 at line 138 column 7
    python_ipc_driver at line 59 column 13
    python_cmd at line 164 column 9
    valid_sym_assumptions at line 38 column 10
    assumptions at line 82 column 7
    syms at line 97 column 13
>>
````

**Comment by cbm755** [comment]
I notice its Python 2.7 but from 2010.  We have tested with 2.7.9.  I'm not sure what the minimum version is...

**Comment by mtmiller** [comment]
The version string `2.7 (r27:82525, Jul  4 2010, 07:43:08) [MSC v.1500 64 bit (AMD64)]` comes from the official Python 2.7.0 installer [python-2.7.amd64.msi](https://www.python.org/download/releases/2.7/), if that helps. Someone should be able to test it and may be able to reproduce. Maybe there is some version at which they fixed some newline problem on Windows.

**Comment by cbm755** [comment]
@mtmiller, yes my thoughts exactly, thanks for tracking it down to a precise version number.  Reading the point-release changelogs might narrow it down too, but no time :(

**Comment by mtmiller** [comment]
I couldn't let this go, so I installed a couple of historical versions of Python. Python 2.7.3 and greater should be ok. Python 2.7.2 and older will hit this bug. This has nothing to do with Windows, I can reproduce on Debian. This has to do with the xml output produced by the minidom `toprettyxml` method.

In modern versions of Python, you get

    <?xml version="1.0" encoding="utf-8"?>
    <output_block>
    <list>
    <item>
    <f>1003</f>
    <f>Communication established.</f>
    </item>
    ...

In Python 2.7.2 and earlier you get this instead

    <?xml version="1.0" encoding="utf-8"?>
    <output_block>
    <list>
    <item>
    <f>
    1003
    </f>
    <f>
    Communication established.
    </f>
    </item>

You could try to make extractblock a little more flexible to match this old output, but it just doesn't seem worth it. Do you even have a place where you check for a minimum Python version yet?

Btw, I just ran the full test suite with Python 2.7.3 and SymPy 1.1.1 with no test failures.

**Comment by cbm755** [comment]
Another heroic bisection by @mtmiller.

> Do you even have a place where you check for a minimum Python version yet?

No, not since we dropped the configure script.  Right place is `private/assert_have_python_and_sympy.m`.

Seems likely there is a minimum Python 3 version too.  We could use Travis to find it I think...  All of this should be an easy fix for someone who wants to get involved!

**Comment by mtmiller** [comment]
Minimum versions are 2.7.3 and 3.2.3. Both 2.7.2 and 3.2.2 suffer from the same `xml.dom.minidom` incompatibility.

It's also worth noting that Python 3.3 was the first to support the `u'foo'` string literal syntax. The SymPy 1.1 release no longer supports Python 3.2. So Python 3.2.3 through 3.2.6 work with octsympy as long as SymPy 1.0 is installed. Python 3.3.0 works with SymPy 1.1.

There is only one small test failure in `python_cmd.m` with Python 3.2.3, where it uses u-string syntax, but other than that all tests pass.

Sadly though, nothing prevents a user from trying to install SymPy 1.1 under Python 3.2. If we want to get real picky, we could throw an error if that known incompatibility is found.

To summarize, something like this

    spver = tuple(map(int, sympy.__version__.split('.')[:2]))
    if sys.version_info < (2, 7, 3):
        raise RuntimeError('Python 2.7.3 or Python 3.2.3 or higher required')
    elif sys.version_info > (3, 0) and sys.version_info < (3, 2, 3):
        raise RuntimeError('Python 2.7.3 or Python 3.2.3 or higher required')
    elif sys.version_info > (3, 0) and sys.version_info < (3, 3) and spver >= (1, 1):
        raise RuntimeError('Python 3.2 is incompatible with SymPy 1.1, please try SymPy 1.0 instead')

**Comment by mtmiller** [comment]
Btw, a nice coincidence is that even if a user is on ancient Ubuntu 12.04 LTS, they have Python 2.7.3 and 3.2.3 out of the box. CentOS 7 users should be ok also. CentOS 6 users (I used to be one, now a vanishing minority) will have to go with anaconda or pyenv or something else.

**Comment by mtmiller** [comment]
I've started working on this, but with a feature test using xml.dom instead of checking version numbers. PR coming soon.
